README.md
# normal

A simple library for working with the normal probability distribution.

## Usage

A simple usage example:

```dart
import 'package:normal/normal.dart';

void main() {
  final d = Normal(),
      xs = [for (var i = 0; i < 14; i++) (i - 6) / 2],
      ps = [for (var i = 1; i < 10; i++) i / 10.0];
      
  // Probability density function.
  print('\nNormal.pdf\n----------\n');
  for (final x in xs) {
    final p = d.pdf(x);
    print(
        '${x.toString().padLeft(4)} |${' ' * (p * 30).round()}:${p.toStringAsFixed(4)}');
  }

  // Cumulative density function.
  print('\nNormal.cdf\n----------\n');
  for (final x in xs) {
    final p = d.cdf(x);
    print(
        '${x.toString().padLeft(4)} |${' ' * (p * 40).round()}:${p.toStringAsFixed(4)}');
  }

  // Quantile function.
  print('\nNormal.inverseCdf\n-----------------\n');
  for (final p in ps) {
    final x = d.quantile(p);
    print('${p.toStringAsFixed(1)}\t${x.toStringAsFixed(4).padLeft(7)}');
  }

  // Generate random numbers.
  print('\nNormal.generate\n---------------\n');
  for (final x in d.generate(10, seed: 0)) {
    print(x.toStringAsFixed(4).padLeft(7));
  }
}
```

Output:

```text
Normal.pdf
----------

-3.0 |:0.0044
-2.5 | :0.0175
-2.0 |  :0.0540
-1.5 |    :0.1295
-1.0 |       :0.2420
-0.5 |           :0.3521
 0.0 |            :0.3989
 0.5 |           :0.3521
 1.0 |       :0.2420
 1.5 |    :0.1295
 2.0 |  :0.0540
 2.5 | :0.0175
 3.0 |:0.0044
 3.5 |:0.0009

Normal.cdf
----------

-3.0 |:0.0013
-2.5 |:0.0062
-2.0 | :0.0228
-1.5 |   :0.0668
-1.0 |      :0.1587
-0.5 |            :0.3085
 0.0 |                    :0.5000
 0.5 |                            :0.6915
 1.0 |                                  :0.8413
 1.5 |                                     :0.9332
 2.0 |                                       :0.9772
 2.5 |                                        :0.9938
 3.0 |                                        :0.9987
 3.5 |                                        :0.9998

Normal.inverseCdf
-----------------

0.1     -1.2817
0.2     -0.8415
0.3     -0.5240
0.4     -0.2529
0.5     -0.0000
0.6      0.2529
0.7      0.5240
0.8      0.8415
0.9      1.2817

Normal.generate
---------------

 0.4679
-0.2936
 0.1079
 0.2815
 0.6585
 0.6758
 0.3304
-1.0116
 0.1754
 1.1171
```

## Features and bugs

Thanks for your interest in *normal*. Please submit [any issues you encounter](https://bitbucket.org/ram6ler/normal/issues).


CHANGELOG.md
## 0.7.0

- Housecleaning.
- Breaking change: switched from abstract to instantiable class.
- Just use one of the Box-Muller pairs during random number generation for decoupling and simpler code.

## 0.6.0

- Null safety; updated style.
- Thanks to Eiko Bleicher.

## 0.5.0

- Initial version


test/normal_test.dart
import "package:normal/normal.dart";

main() {
  // Probability density function.
  print("\nNormal.pdf\n----------\n");
  for (final x in List.generate(14, (i) => (i - 6) / 2)) {
    final p = Normal.pdf(x);
    print(
        "${x.toString().padLeft(4)} |${" " * (p * 30).round()}:${p.toStringAsFixed(4)}");
  }

  // Cumulative density function.
  print("\nNormal.cdf\n----------\n");
  for (final x in List.generate(14, (i) => (i - 6) / 2)) {
    final p = Normal.cdf(x);
    print(
        "${x.toString().padLeft(4)} |${" " * (p * 40).round()}:${p.toStringAsFixed(4)}");
  }

  // Quantile function.
  print("\nNormal.inverseCdf\n-----------------\n");
  for (final p in List.generate(9, (i) => (i + 1) * 0.1)) {
    final x = Normal.inverseCdf(p);
    print("${p.toStringAsFixed(1)}\t${x.toStringAsFixed(4).padLeft(7)}");
  }

  // Generate random numbers.
  print("\nNormal.generate\n---------------\n");
  for (final x in Normal.generate(10, seed: 0)) {
    print(x.toStringAsFixed(4).padLeft(7));
  }
}


lib/src/normal_base.dart
import 'dart:math' as math;

const _p = 0.2316419,
    _nb = 5,
    _b = [
      0.319381530,
      -0.356563782,
      1.781477937,
      -1.821255978,
      1.330274429,
    ],
    _nc = 3,
    _c = [2.515517, 0.802853, 0.010328],
    _nd = 4,
    _d = [1.0, 1.432788, 0.189269, 0.001308];

double _standardDensity(num x) => math.exp(-x * x / 2) / math.sqrt(2 * math.pi);

class Normal {
  Normal([this.mean = 0.0, this.variance = 1.0])
      : standardDeviation = math.sqrt(variance);

  /// The mean of the normal distribution.
  final num mean;

  /// The variance of the normal distribution.
  final num variance;

  /// The standard deviation of the normal distribution.
  final num standardDeviation;

  /// Gives the normal probability density at [x].
  num pdf(num x) =>
      _standardDensity((x - mean) / standardDeviation) / standardDeviation;

  /// Gives the  normal cumulative distribution function.
  ///
  /// Approximation within 7.5E-8 (*Abramowitz* 26.2.17)
  /// of the cumulative probability.
  ///
  num cdf(num x) {
    final z0 = (x - mean) / standardDeviation,
        z = z0.abs(),
        t = 1 / (1 + _p * z),
        tps = [for (var i = 0; i < _nb; i++) math.pow(t, i + 1)],
        c = _standardDensity(z) *
            [for (var i = 0; i < _nb; i++) _b[i] * tps[i]]
                .fold(0, (a, b) => a + b);
    return z0 < 0 ? c : 1 - c;
  }

  /// Gives the quantile associated with an input probability.
  ///
  /// Approximation within 4.5E-4 (*Abramowitz* 26.2.23) of the normal
  /// quantile associated with [p].
  ///
  num quantile(num p) {
    final standardDeviation = math.sqrt(variance),
        restrictedP = p > 0.5 ? 1 - p : p,
        t = math.sqrt(math.log(1 / restrictedP / restrictedP)),
        tps = [for (var i = 0; i < _nd; i++) math.pow(t, i)],
        cs = [for (var i = 0; i < _nc; i++) _c[i] * tps[i]]
            .fold(0.0, (a, b) => a + b),
        ds = [for (var i = 0; i < _nd; i++) _d[i] * tps[i]]
            .fold(0.0, (a, b) => a + b),
        z = t - cs / ds;

    return (p < 0.5 ? -1 : 1) * (mean + z * standardDeviation);
  }

  /// Generates random data drawn from a normal distribution.
  ///
  /// Produces a pseudorandom sample drawn from a normal distribution
  /// through the Box-Muller algorithm.
  ///
  List<num> generate(int n, {int? seed}) {
    final rand = seed == null ? math.Random() : math.Random(seed),
        standardDeviation = math.sqrt(variance);
    num next() {
      final u1 = rand.nextDouble(),
          u2 = rand.nextDouble(),
          r = math.sqrt(-2 * math.log(u1)),
          t = 2 * math.pi * u2;
      return r * math.cos(t);
    }

    final zScores = [for (var _ = 0; _ < n; _++) next()];

    return [for (final z in zScores) z * standardDeviation + mean];
  }
}


example/normal_example.dart
import 'package:normal/normal.dart';

void main() {
  final d = Normal(),
      xs = [for (var i = 0; i < 14; i++) (i - 6) / 2],
      ps = [for (var i = 1; i < 10; i++) i / 10.0];

  // Probability density function.
  print('\nNormal.pdf\n----------\n');
  for (final x in xs) {
    final p = d.pdf(x);
    print(
        '${x.toString().padLeft(4)} |${' ' * (p * 30).round()}:${p.toStringAsFixed(4)}');
  }

  // Cumulative density function.
  print('\nNormal.cdf\n----------\n');
  for (final x in xs) {
    final p = d.cdf(x);
    print(
        '${x.toString().padLeft(4)} |${' ' * (p * 40).round()}:${p.toStringAsFixed(4)}');
  }

  // Quantile function.
  print('\nNormal.inverseCdf\n-----------------\n');
  for (final p in ps) {
    final x = d.quantile(p);
    print('${p.toStringAsFixed(1)}\t${x.toStringAsFixed(4).padLeft(7)}');
  }

  // Generate random numbers.
  print('\nNormal.generate\n---------------\n');
  for (final x in d.generate(10, seed: 0)) {
    print(x.toStringAsFixed(4).padLeft(7));
  }
}
